api/projects/tags/migrations/0005_add_tag_fields_for_stale_flags_logic.py

```diff
@@ -0,0 +1,75 @@
+# Generated by Django 3.2.23 on 2023-12-06 15:11
+import re
+
+from django.apps.registry import Apps
+from django.db import migrations, models
+from django.db.backends.base.schema import BaseDatabaseSchemaEditor
+
+
+PROTECTED_LABELS = {"protected", "donotdelete", "permanent"}
+LABEL_REGEX = re.compile(r"[ _]")
+
+
+def get_sanitised_label(label: str) -> str:
+    return LABEL_REGEX.sub("", label.lower())
+
+
+def mark_existing_tags_as_permanent(
+    apps: Apps, schema_editor: BaseDatabaseSchemaEditor
+) -> None:
+    """
+    Update all tags in the database which the FE treats as 'permanent' through regex matching
+    to have the new is_permanent label. The FE will subsequently be updated to use the
+    is_permanent attribute instead of regex matching.
+    """
+    tag_class = apps.get_model("tags", "tag")
+
+    to_update = []
+
+    for tag in filter(
+        lambda t: get_sanitised_label(t.label) in PROTECTED_LABELS,
+        tag_class.objects.all(),
+    ):
+        tag.is_permanent = True
+        to_update.append(tag)
```

Comment:
Do we have an approximation of how long will it run for SaaS?

Reply:
There are only ~3000 tags in our production database in total. We'll have to iterate over all of those and run the regex. Of those that match the regex, there are only 62. 

Query to get this data: 

```sql
select count(*)
from tags_tag t
where replace(replace(lower("label"), ' ', ''), '_', '') in ('protected', 'donotdelete', 'permanent');
```

Here's a simple script that attempts to replicate it: 

```python
import random
import re
import time
from dataclasses import dataclass

PROTECTED_LABELS = {"protected", "donotdelete", "permanent"}
LABEL_REGEX = re.compile(r"[ _]")


def get_sanitised_label(label: str) -> str:
    return LABEL_REGEX.sub("", label.lower())


@dataclass
class Dummy:
    """Dummy dataclass to at least attempt to simulate model serialization"""

    label: str


matching_labels = [
    random.choice(
        [
            "do not delete",
            "protected",
            "Protected",
            "do_not_delete",
            "permanent",
            "Permanent",
        ]
    )
    for _ in range(65)
]
non_matching_labels = ["not matching" * 3000]
labels = [*matching_labels, *non_matching_labels]


if __name__ == "__main__":
    start = time.time()
    print(f"Starting at: {start}")

    to_update = []

    for label in filter(
        lambda l: get_sanitised_label(l) in PROTECTED_LABELS,
        labels,
    ):
        to_update.append(label)

    end = time.time()
    print(f"Finished at: {end}")

    print(f"Process took {end - start}")
```

And the output: 

```
Starting at: 1707500632.317796
Finished at: 1707500632.318275
Process took 0.00047898292541503906
```

tl;dr

>Do we have an approximation of how long will it run for SaaS?

Probably a few ms.